store/sqlstore/plugin_store.go

```diff
@@ -71,41 +71,34 @@ func (ps SqlPluginStore) SaveOrUpdate(kv *model.PluginKeyValue) store.StoreChann
 	})
 }
 
-func (ps SqlPluginStore) CompareAndUpdate(kv *model.PluginKeyValue, oldValue []byte) store.StoreChannel {
-	return store.Do(func(result *store.StoreResult) {
-		if result.Err = kv.IsValid(); result.Err != nil {
-			return
-		}
+func (ps SqlPluginStore) CompareAndUpdate(kv *model.PluginKeyValue, oldValue []byte) (bool, *model.AppError) {
+	if err := kv.IsValid(); err != nil {
+		return false, err
+	}
 
-		if ps.DriverName() == model.DATABASE_DRIVER_POSTGRES || ps.DriverName() == model.DATABASE_DRIVER_MYSQL {
-			updateResult, err := ps.GetMaster().Exec(
-				`UPDATE PluginKeyValueStore SET PValue = :New WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Old`,
-				map[string]interface{}{
-					"PluginId": kv.PluginId,
-					"Key":      kv.Key,
-					"Old":      oldValue,
-					"New":      kv.Value,
-				})
-			if err != nil {
-				result.Err = model.NewAppError("SqlPluginStore.CompareAndUpdate", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
-				return
-			}
+	updateResult, err := ps.GetMaster().Exec(
+		`UPDATE PluginKeyValueStore SET PValue = :New WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Old`,
```

Comment:
Sorry for the late thought here, but what happens the /first/ time I want to record this value? I can't use `CompareAndUpdate` to `INSERT`, so I guess I start with a `SaveOrUpdate`? Will two plugins racing on `SaveOrUpdate` fail to notice each other's clobbering values?

In theory, could we accept an `oldValue` of `nil`, and update this code to do an `INSERT` instead? The `INSERT` would fail for the /other/ plugin, which could then do a `CompareAndUpdate` after fetching the new value.

Reply:
Fair point. Initially I was thinking even if there were race conditions for the initial insert it would be ok since it's a very tiny window, but it does make sense that this does the initial set if there is no key. So this this method becomes more of a `CompareAndSet`
1. Attempt to `INSERT` when `oldValue` is `nil`
2. `UPDATE` otherwise.

Reply:
Good thought. With `INSERT` now supported, `KVCompareAndSet` api has complete basic transactional support :)